com.unity.shadergraph/Editor/Generation/Processors/Generator.cs

```diff
@@ -13,56 +14,98 @@
 
 namespace UnityEditor.ShaderGraph
 {
+    struct GeneratedShader
+    {
+        public string codeString;
+        public string shaderName;
+        public List<PropertyCollector.TextureInfo> assignedTextures;
+        public bool isErrorShader;
+    }
+
     class Generator
     {
         const string kDebugSymbol = "SHADERGRAPH_DEBUG";
 
-        GraphData m_GraphData;
-        AbstractMaterialNode m_OutputNode;
-        Target[] m_Targets;
-        List<BlockNode> m_ActiveBlocks;
-        List<BlockNode> m_TemporaryBlocks;
-        GenerationMode m_Mode;
-        string m_Name;
-
-        ShaderStringBuilder m_Builder;
-        List<PropertyCollector.TextureInfo> m_ConfiguredTextures;
-        AssetCollection m_assetCollection;
-        bool m_humanReadable;
-
-        public string generatedShader => m_Builder.ToCodeBlock();
-        public List<PropertyCollector.TextureInfo> configuredTextures => m_ConfiguredTextures;
-        public List<BlockNode> temporaryBlocks => m_TemporaryBlocks;
-
-        public Generator(GraphData graphData, AbstractMaterialNode outputNode, GenerationMode mode, string name, AssetCollection assetCollection, bool humanReadable = false)
+        // readonly data setup in constructor
+        readonly GraphData m_GraphData;
+        readonly AbstractMaterialNode m_OutputNode;
+        readonly GenerationMode m_Mode;
+        readonly string m_PrimaryShaderFullName;
+        readonly AssetCollection m_AssetCollection;
+        readonly bool m_HumanReadable;
+        readonly ReadOnlyCollection<BlockNode> m_ActiveBlocks;
+        readonly ReadOnlyCollection<Target> m_Targets;
+        readonly ReadOnlyCollection<TargetSetupContext> m_TargetContexts;
+        readonly ReadOnlyCollection<string> m_AdditionalShaderIDs;
+
+        readonly GeneratedShader m_PrimaryShader;
+        readonly List<BlockNode> m_PrimaryShaderTemporaryBlocks;
+
+        // direct accessors for primary shader results
+        public string generatedShader => m_PrimaryShader.codeString;
+        public List<PropertyCollector.TextureInfo> configuredTextures => m_PrimaryShader.assignedTextures;
+        public List<BlockNode> temporaryBlocks => m_PrimaryShaderTemporaryBlocks;
+
+        // accessor for all generated shaders
+        public IEnumerable<GeneratedShader> allGeneratedShaders
```

Comment:
Yay for comments and more type safety :-)

Reply:
Yeah, ensuring that I didn't miss any data that gets mutated during the generation..   Still possible I guess it can mutate data within one of the classes here, but it's a bit safer now.